File: gospider/antibot/night_mode.go
package antibot

import "time"

// NightModePolicy lowers crawl aggressiveness during configured quiet hours.
type NightModePolicy struct {
	Enabled         bool
	StartHour       int
	EndHour         int
	DelayMultiplier float64
	RateLimitFactor float64
}

// DefaultNightModePolicy mirrors the shared framework defaults used by pyspider.
func DefaultNightModePolicy() NightModePolicy {
	return NightModePolicy{
		Enabled:         true,
		StartHour:       23,
		EndHour:         6,
		DelayMultiplier: 1.5,
		RateLimitFactor: 0.5,
	}
}

// IsActive reports whether the provided time falls into the configured quiet-hour window.
func (p NightModePolicy) IsActive(at time.Time) bool {
	if !p.Enabled {
		return false
	}
	hour := at.Hour()
	if p.StartHour == p.EndHour {
		return true
	}
	if p.StartHour < p.EndHour {
		return hour >= p.StartHour && hour < p.EndHour
	}
	return hour >= p.StartHour || hour < p.EndHour
}

// ApplyDelay scales a delay when the policy is active.
func (p NightModePolicy) ApplyDelay(base time.Duration, at time.Time) time.Duration {
	if base <= 0 || !p.IsActive(at) {
		return base
	}
	adjusted := time.Duration(float64(base) * p.DelayMultiplier)
	if adjusted < base {
		return base
	}
	return adjusted
}

// ApplyRateLimit scales a rate limit when the policy is active.
func (p NightModePolicy) ApplyRateLimit(base float64, at time.Time) float64 {
	if base <= 0 || !p.IsActive(at) {
		return base
	}
	adjusted := base * p.RateLimitFactor
	if adjusted <= 0 {
		return base
	}
	return adjusted
}

File: rustspider/src/antibot/night_mode.rs
use chrono::{DateTime, Timelike};
use serde::{Deserialize, Serialize};
use std::time::Duration;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct NightModePolicy {
    pub enabled: bool,
    pub start_hour: u8,
    pub end_hour: u8,
    pub delay_multiplier: f64,
    pub rate_limit_factor: f64,
}

impl Default for NightModePolicy {
    fn default() -> Self {
        Self {
            enabled: true,
            start_hour: 23,
            end_hour: 6,
            delay_multiplier: 1.5,
            rate_limit_factor: 0.5,
        }
    }
}

impl NightModePolicy {
    pub fn is_active<Tz>(&self, at: DateTime<Tz>) -> bool
    where
        Tz: chrono::TimeZone,
    {
        if !self.enabled {
            return false;
        }
        let hour = at.hour() as u8;
        if self.start_hour == self.end_hour {
            return true;
        }
        if self.start_hour < self.end_hour {
            hour >= self.start_hour && hour < self.end_hour
        } else {
            hour >= self.start_hour || hour < self.end_hour
        }
    }

    pub fn apply_delay<Tz>(&self, base: Duration, at: DateTime<Tz>) -> Duration
    where
        Tz: chrono::TimeZone,
    {
        if base.is_zero() || !self.is_active(at) {
            return base;
        }
        let adjusted = Duration::from_secs_f64(base.as_secs_f64() * self.delay_multiplier);
        if adjusted < base {
            base
        } else {
            adjusted
        }
    }

    pub fn apply_rate_limit<Tz>(&self, base: f64, at: DateTime<Tz>) -> f64
    where
        Tz: chrono::TimeZone,
    {
        if base <= 0.0 || !self.is_active(at) {
            return base;
        }
        let adjusted = base * self.rate_limit_factor;
        if adjusted <= 0.0 {
            base
        } else {
            adjusted
        }
    }
}
